File: Service/Account/middleware/authentication.go
package middleware

import (
	"github.com/labstack/echo"
	"go-module/libs/custom_type" 
	"go-module/libs/jwt"
	"net/http"
	"strings"

	"fmt"
)

func Authentication(next echo.HandlerFunc) echo.HandlerFunc {

	return func(c echo.Context) error {		
		// Http error response 
		httpErr := new(echo.HTTPError)

		// Check if account authorized
		request := c.Request()
		authorHeader := request.Header.Get("Authorization")
		splitted := strings.Split(authorHeader, " ")
		if (authorHeader == "" || len(splitted) != 2) {
			httpErr.Code = http.StatusForbidden
			httpErr.Message = "Unauthorized."
			return httpErr
		}

		// Verify access token
		tokenPart := splitted[1]
		token, err := jwt.Verify(tokenPart)
		
		fmt.Println("err: ", err)
		if (err != nil || !token.Valid) {
			httpErr.Code =  http.StatusInternalServerError
			httpErr.Message = "Invalid access token."
			return httpErr
		}

		// Get claims from access token
		claims := token.Claims.(*custom_type.Claims)
		
		// Forwarding claim info to the next middlware
		authenInfo := custom_type.AuthenInfo{}
		authenInfo = claims.AuthenInfo
		c.Set("Role", authenInfo.Role)

		// Go to next 
		err = next(c)
		return err
	}
}

File: Service/Product/src/middlewares/authentication.ts
'use strict'
import jwt from 'jsonwebtoken';
import { Response, NextFunction } from 'express';
import { ACCESS_TOKEN_SECRET } from '../config.json';

import IGetUserAuthInfoRequest 
    from '../libs/interfaces/auth-info-request';

const userAuthentication = async (req: IGetUserAuthInfoRequest, 
        res: Response, next: NextFunction) => {
    try {
        // Check if account authorized
        const author_header = req.headers?.authorization || '';
        let statusCode: number, message: string;
        const splitted = author_header.split(' ');
        if (!author_header || splitted.length != 2) {
            statusCode = 403; // Forbidden
            message = 'Unauthorized.';
            return res.status(statusCode).json({ 
                user: null, message
            })
        }

        // Verify access token
        const accessToken = splitted[1] || '';
        jwt.verify(accessToken, ACCESS_TOKEN_SECRET, (err, 
                decodedToken: jwt.JwtPayload) => {

            // Verify failed.
            if (err) {
                statusCode = 500; // Internal Server Error
                message = 'Access Token is not valid!';
                return res.status(500).json({ 
                    user: null, message 
                });
            }       

            // Verify passed.
            const { email = '', name = '', role = '' } = decodedToken;
            req.user = { email, name, role };
            next()
        })
        
    } catch (err) {
        return res.status(403).json({ 
            user: null, message: 'Invalid access token.' 
        });   
    }
}

export default { userAuthentication };
